### PR TITLE
Integrated xcm simulator and XcmpHandler

### PR DESCRIPTION
**Pull request by zktony**
Integrated xcm simulator and XcmpHandler

### CODE REVIEW

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

File: pallets/xcm-handler/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+homepage = "https://substrate.io"
+repository = "https://github.com/paritytech/substrate/"
+edition = "2021"
+
+[package.metadata.docs.rs]
+targets = ["x86_64-unknown-linux-gnu"]
+
+[dependencies]
+codec = { package = "parity-scale-codec", version = "3.0.0", features = ["derive"], default-features = false }
+scale-info = { version = "2.2.0", default-features = false, features = ["derive"] }
```
[the remark this answers is not in the document]
Comment (zktony) [reply pallets/xcm-handler/Cargo.toml:16]: done

File: pallets/xcm-handler/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,41 @@
+[package]
+name = "xcm-handler"
+authors = ["Anonymous"]
+description = "FRAME pallet template for defining custom runtime logic."
```
[the remark this answers is not in the document]
Comment (zktony) [reply pallets/xcm-handler/Cargo.toml:4]: done

File: pallets/xcm-handler/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+targets = ["x86_64-unknown-linux-gnu"]
+
+[dependencies]
+codec = { package = "parity-scale-codec", version = "3.0.0", features = ["derive"], default-features = false }
+scale-info = { version = "2.2.0", default-features = false, features = ["derive"] }
+
+# Substrate
+frame-benchmarking = { git = "https://github.com/paritytech/substrate", default-features = false, optional = true, branch = "polkadot-v0.9.32" }
+frame-support = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.32" }
+frame-system = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.32" }
```
[the remark this answers is not in the document]
Comment (zktony) [reply pallets/xcm-handler/Cargo.toml:21]: done

File: runtime/src/xcm_config.rs [hunk 4]
```diff
@@ -147,8 +182,33 @@ impl ShouldExecute for DenyReserveTransferToRelayChain {
...
+				DepositAsset { beneficiary, .. } => {
+					if let Some(recipient) = DepositEvent::get_recipient(beneficiary) {
+						deposit_event.set_recipient(recipient);
+					}
+				}
+				_ => {}
+			};
+
+		}
+		// TODO: Assets Pallet and AssetId convertor is not implemented yet. Issue :#35
```
[the remark this answers is not in the document]
Comment (zktony) [reply runtime/src/xcm_config.rs:204]: done

File: pallets/xcm-handler/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,105 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+/// Edit this file to define custom logic or remove it if it is not needed.
+/// Learn more about FRAME and the core library of Substrate FRAME pallets:
+/// <https://docs.substrate.io/v3/runtime/frame>
```
[the remark this answers is not in the document]
Comment (zktony) [reply pallets/xcm-handler/src/lib.rs:5]: done

File: pallets/xcm-handler/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+		pub fn deposit_asset(origin: OriginFor<T>, recipient: T::AccountId, asset_id: u128, amount: u128) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+			<IngressMessages<T>>::try_mutate(|ingress_messages| {
+				ingress_messages
+					.try_push(TheaMessage::AssetDeposited(recipient.clone(), asset_id, amount))
+					.map_err(|_| Error::<T>::IngressMessagesLimitReached)?;
+				Self::deposit_event(Event::AssetDeposited(recipient, asset_id, amount));
+				Ok(().into())
+			})
+		}
```
[the remark this answers is not in the document]
Comment (zktony) [reply pallets/xcm-handler/src/lib.rs:97]: yeah .. I moved it

File: pallets/xcm-handler/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+/// <https://docs.substrate.io/v3/runtime/frame>
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::{dispatch::DispatchResultWithPostInfo, pallet_prelude::*, PalletId};
+	use frame_system::pallet_prelude::*;
+	use frame_support::sp_runtime::traits::AccountIdConversion;
+
+	//TODO Replace this with TheaMessages #Issue:
```
[the remark this answers is not in the document]
Comment (zktony) [reply pallets/xcm-handler/src/lib.rs:14]: done

File: pallets/xcm-handler/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+	// Dispatchable functions allows users to interact with the pallet and invoke state changes.
+	// These functions materialize as "extrinsics", which are often compared to transactions.
+	// Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+
+		///Deposit to Orderbook
+		#[pallet::weight(Weight::from_ref_time(10_000) + T::DbWeight::get().writes(1))]
+		pub fn deposit_asset(origin: OriginFor<T>, recipient: T::AccountId, asset_id: u128, amount: u128) -> DispatchResultWithPostInfo {
```
[the remark this answers is not in the document]
Comment (zktony) [reply pallets/xcm-handler/src/lib.rs:88]: Hey, it is just designed for Milestone 1. I have to add more things after integrating the asset transactor. I will add docs and test cases after that.

**Review by serhii-temchenko — APPROVED** [review]

**Review by Gauthamastro — APPROVED** [review]

**Pull request merged**